### ISSUE TITLE
Wrong size of legend symbols when using Geometry generator symbology

### ISSUE THREAD

**Issue by gwenandres** [qgis/QGIS#33755]
Hello, 
I get really big symbols in the legend on a point layer with surface or line geometry generator when I choose categorized symbols.

For example : 
- I create a new point memory layer 
- I add a column "type"  (integer)
- For style, I choose categorized symbols on "type" column
- For symbols, I choose surface geometry generator (`buffer ($ geometry, 2)`)
- Then in legend I get really big symbols for each type.

I am working with QGIS 3.4.14 LTR on Windows 7 or 10.
I try with QGIS 3.10.1, I have the same behaviour.
![Qgis_issue_categorized_symbols](https://user-images.githubusercontent.com/15909387/72250057-0c70bd80-35fb-11ea-9be8-b7ab989b59ff.png)

**Comment by gioman** [comment]
@gwenandres your layer has a CRS that has degrees as units?

**Comment by gwenandres** [comment]
The CRS is 2154 so it has meters as units.

**Comment by gioman** [comment]
@gwenandres Can't confirm here on 3.10.1 on Ubuntu 18.04 but I'm not totally sure I followed the exact steps that lead you to see this problem.

![Screenshot_20200116_134557](https://user-images.githubusercontent.com/1951107/72530155-a7b99b00-3866-11ea-81f7-3689a1587ce8.png)

**Comment by gwenandres** [comment]
@gioman  Sorry if I am not clear. I have the problem on a point layer, not on a surface layer. 
![QGIS3-pointlayer](https://user-images.githubusercontent.com/15909387/72532397-a2168200-3873-11ea-9f9a-f32a62183c96.png)

Yes it can be strange but I have point layer with surface generator.

**Comment by gioman** [comment]
> I have the problem on a point layer

@gwenandres oh I see... moreover not only the legend symbol does not make sense, but also the generated area on map is not visible at all.

**Comment by SrNetoChan** [comment]
I was able to replicate this. It only happens on Rule-based (maybe categorized too) Symbol, but not in single symbol, where it works just fine.

**Comment by jonnyforestGIS** [comment]
@SrNetoChan I confirm as well this behaviour when using the geometry generator and happens with categorized symbol. Cheers

**Comment by jonnyforestGIS** [comment]
@gioman having this #39777 as a possible resolution for this issue, this ticket can be close?

**Comment by gioman** [comment]
> @gioman having this #39777 as a possible resolution for this issue, this ticket can be close?

@jonnyforestGIS have you tested the fix using the scenario described in this ticket?

**Comment by jonnyforestGIS** [comment]
@gioman you're right. 
I did a tested with the master and change the options but the behaviour with the new options but the issue still there. Please see my comment in #39777 , I will wait for @nirvn to answer, maybe I'm doing something wrong. Cheers

**Comment by SrNetoChan** [comment]
I have confirmed this on master. @nirv fix does not fix the issue when the geometry generator is uses in categorised, graduated or rule based renderers. But works like a charm for single symbol renderer

**Comment by gioman** [comment]
See also https://github.com/qgis/QGIS/issues/29119

**Comment by SrNetoChan** [comment]
I guess we can close one as duplicate.

**Comment by gioman** [comment]
> I guess we can close one as duplicate.

already did :)

**Comment by sickel** [comment]
I can confirm the same behaviour in 3.20.3. Using a postGIS point layer with a generated polygon around the point. It is a bit unstable, in some cases I see the (at least by me) expected behaviour with a square a little bit bigger than the point symbol, in other cases the height is as expected (althought the top and bottom is missing, but the symbol is approx 10cm wide, quite often, it is a box of about 10x10cm2. The dataset is in WGS84. The generated geometry is made by another geometry in the the same postGIS table.
![Skjermbilde](https://user-images.githubusercontent.com/3080462/135217694-1d40d854-5809-43c4-9021-6c8ffd341a97.PNG)

**Comment by sickel** [comment]
This is another dataset made in the same way in the same project. Also WGS84. Only difference is that it is a graduated symbology whereas the first was rule base
![Skjermbilde](https://user-images.githubusercontent.com/3080462/135221292-965d6dcc-d89c-4699-997a-670175fe187f.PNG)
d.

**Comment by sickel** [comment]
To follow up on my previous comment, this is what the legend looked like when I just opened the same project. The height seems to be cropped, the symbols are too wide,
![Skjermbilde](https://user-images.githubusercontent.com/3080462/135587496-0c9a0d7f-29de-4a9b-91d9-fd0024f562f1.PNG)

**Comment by Pedro-Murteira** [comment]
I can still replicate this on QGIS 3.16.14 and 3.22.1.

**Comment by SrNetoChan** [comment]
There is an option to limit the size of the legend patches in the layers panel. It seems to work fine for symbols where the size is controlled by an expression, but not by geometry generators.

In this project, this is clear
https://mega.nz/file/5MFkERjQ#OXEO-g6oUypng9eMvx1KDHA4lRmMbC8gKN40gz6dMcc

**Comment by martyvis** [comment]
Looks like there is a general problem with Geometry Generator scaling I am seeing as well. I have a Point layer and I add a simple square using Geometry Generator expression as follows 👎 
`make_line($geometry, translate($geometry,0,10.0), translate($geometry,10.0,10.0), translate($geometry,10.0,0.0), $geometry)`

With Map Units set and using a CSR of EPSG:3857 (WGS84/Pseudo Mercator) you would expect the square to be 10m by 10m. Instead it looks like it is maybe 10 degrees x 10 degrees?
![Screenshot 2022-06-27 191432](https://user-images.githubusercontent.com/7751962/175904884-e3bb570c-c874-4d76-8c34-d2115b7d310c.png)
![Screenshot 2022-06-27 191355](https://user-images.githubusercontent.com/7751962/175904901-7e694af9-d96d-4485-b7c7-5b070e355059.png)

Running 3.24.2-Tisler on Windows 11

**Comment by nyalldawson** [comment]
@martyvis 

Is your layer in 4326? The geometry generator map units refer to the layer's units.

**Comment by martyvis** [comment]
@nyalldawson  Yes it was. I think I have worked out I had a workflow issue. When I create a new layer, usually a scratch layer I change its Geometry Type (say to Points) and add fields if needed but don't change the CRS. Somehow I thought the scratch layer would have followed the map layer I had underneath. I worked out if I change the CRS (eitger on creation or afterwards, it all works). The thing I still notice though is they if you open Geometry Generator expression build and look at the "map_units" variable it shows it as being "meters". So obviously this isn't correctly reflecting the Map Units that the layer is using. (Also if you have the layer selected and move your mouse the coordinates in the bottom still show metres and not degrees.

**Comment by AlisterH** [comment]
> Still happening in 3.36.1
> 
> I have this symbol, everything works fine

It is smaller than it is in the map.  I guess in a case like yours this is desirable, but I would usually want it to be the same size.
 
> If I use the same one, but categorized, I get this

Bigger than in the map.  Again, I would normally want it the same size, but you apparently want it smaller.
Ideally it would be configurable, but it should at least be consistent between the different 

I see the same behaviour even though I don't use a geometry generator - are you sure this ticket is actually relevant?
Note that I'm using sizes in pixels or mm, not map units.

**Comment by arkanoid87** [comment]
still happening in 3.40.9-Bratislava

**Comment by gioman** [comment]
> still happening in 3.40.9-Bratislava

Should we reopen this ticket? Or better to ask to file a new one?
